components/script/script_thread.rs

```diff
@@ -1471,28 +1476,292 @@ impl ScriptThread {
+                while let Some((size, size_type)) = pending_resize_events.pop_front() {
+                    // Resize steps.
+                    self.run_the_resize_steps(pipeline_id, size, size_type);
+
+                    // Evaluate media queries and report changes.
+                    document
+                        .window()
+                        .evaluate_media_queries_and_report_changes();
+
+                    // https://html.spec.whatwg.org/multipage/#img-environment-changes
+                    // As per the spec, this can be run at any time.
+                    document.react_to_environment_changes()
+                }
+
+                // Update animations and send events.
+                self.update_animations_and_send_events();
+
+                // TODO: fullscreen steps(#31866).
+
+                // TODO: context lost steps(#31868).
+
+                // Run the animation frame callbacks.
+                self.handle_tick_all_animations(pipeline_id);
+
+                // TODO: resize observer steps(#31006).
+
+                // TODO: if the focused area of doc is not a focusable area,
+                // then run the focusing steps for doc's viewport.
+                // (#31870)
+
+                // TODO: perform pending transition operations(https://drafts.csswg.org/css-view-transitions/).
+
+                // TODO: run the update intersection observations steps(#31021).
+
+                // TODO: mark paint timing(https://w3c.github.io/paint-timing).
 
-        for (id, document) in self.documents.borrow().iter() {
-            // Only process a resize if layout is idle.
-            if let Some((size, size_type)) = document.window().steal_resize_event() {
-                resizes.push((id, size, size_type));
+                // TODO: Update the rendering: consolidate all reflow calls into one here?(#31871)
+
+                // TODO: process top layer removals(https://drafts.csswg.org/css-position-4/#process-top-layer-removals)
             }
         }
+    }
 
-        for (id, size, size_type) in resizes {
-            self.handle_event(id, ResizeEvent(size, size_type));
+    /// <https://html.spec.whatwg.org/multipage/#event-loop-processing-model:rendering-opportunity>
+    fn rendering_opportunity(&self, pipeline_id: PipelineId) {
+        *self.last_render_opportunity_time.borrow_mut() = time::precise_time_ns();
+
+        // Note: the pipeline should be a navigable with a rendering opportunity,
+        // and we should use this opportunity to queue one task for each navigable with
+        // an opportunity in this script-thread.
+        let document = match self.documents.borrow().find_document(pipeline_id) {
+            Some(document) => document,
+            None => {
+                return warn!(
+                    "Trying to update the rendering for closed pipeline {}.",
+                    pipeline_id
+                )
+            },
+        };
```

Comment:
Again, I would use the pattern described above and avoid `return` with statements that return void as it introduces clippy warnings.